containerd-shim-spin-config.md

```diff
@@ -0,0 +1,373 @@
+extended to support configuring the platform and application execution,
+respectively. Users should be able to configure Spin resources and executions
+side-by-side in the `SpinApp` CRD. On the other hand, platform configuration is
+not done on a per application basis. It could be configured in the
+`SpinAppExecutor` CRD. This would help users easily see how the platform is
+configured for the executor their application is using. It would also enable
+creating multiple executors with distinct configuration.
+
+### Spin Application Execution Configuration
+
+All Spin application execution configuration will be defined as known
+environment variables with an expected range of values. All variables must have
+a `SPIN` prefix. Each executor will define which environment variables it
+supports. The following is a subset of configuration values that may be
+supported:
+
+| Key | Spin CLI |  Example Value|
+| ---- | ---- | ---- |
+| SPIN_HTTP_LISTEN_ADDR | `spin up --listen` | "0.0.0.0:3000" |
+| SPIN_OTEL_EXPORTER_OTLP_ENDPOINT | `OTEL_EXPORTER_OTLP_ENDPOINT=http://123.4.5.6:4318 spin up` | "http://123.4.5.6:4318" |
+| SPIN_LOG_DIR | `spin up --log-dir /tmp/log` | "/tmp/log" |
+| SPIN_RUNTIME_CONFIG_FILE | `spin up --runtime-config-file /var/config.toml` | "/var/config.toml" |
+| SPIN_AWS_DEFAULT_REGION | `AWS_DEFAULT_REGION="us-west-2" spin up` | "us-west-2" |
+| SPIN_AWS_ACCESS_KEY_ID | `AWS_ACCESS_KEY_ID="ABC" spin up` | "ABC" |
+| SPIN_AWS_SECRET_ACCESS_KEY | `AWS_SECRET_ACCESS_KEY="123" spin up` | "123" |
+| SPIN_AWS_SESSION_TOKEN | `AWS_SESSION_TOKEN="token" spin up` | "token" |
+
+These values will be configured as environment variables in the Container spec,
+and will be filtered out from all other Container environment variables. All
+environment variables that do not have the `SPIN` prefix, will be set in the
+environment of the Wasm component instances before execution. The fact that some
+container environment variables are configuring the runtime rather than being
+set in the application will feel strange for users who are used to configuring
+application environment variables in Linux containers. Unfortunately, as
+explained in the [alternatives section](#alternatives-considered), there is no
+other smooth path to configuring execution at the application level.
+
+> Note: Variables prefixed with `SPIN_VARIABLE` will be set as [Spin application
+> variables](https://developer.fermyon.com/spin/v2/variables#application-variables)
+> via the Spin environment variable provider as is the experience with the Spin
+> CLI.
+
+Each of the Spin application execution environment variables should be
+configurable from the SpinApp CRD under `SpinApp.spec.runtimeEnvironment`
+
+```sh
+apiVersion: core.spinoperator.dev/v1alpha1
+kind: SpinApp
+metadata:
+  name: myspinapp
+spec:
+  image: "ghcr.io/myuser:v1"
+  replicas: 1
+  executor: containerd-shim-spin
+  runtimeEnvironment:
+  - name: SPIN_HTTP_LISTEN_ADDR
+    value: "0.0.0.0:3000"
+```
```

Comment:
If I interpret this design correctly, then there is no way to set `OTEL_EXPORTER_OTLP_ENDPOINT` on a `SpinAppExecutor` such that it would apply to all `SpinApp`s using that executor. Is that right?

One other concern I have is that the model of setting `OTEL_EXPORTER_OTLP_ENDPOINT` on the `SpinApp` does it work very well with any executors that aren't ephemeral like the shim. Some executors might not be able to change there exporter endpoint on the fly like the shim can.

Reply:
> If I interpret this design correctly, then there is no way to set OTEL_EXPORTER_OTLP_ENDPOINT on a SpinAppExecutor such that it would apply to all SpinApps using that executor. Is that right?

Not necessarily, we can update the SpinAppExecutor to support setting universal `runtimeEnvironment` variables


> One other concern I have is that the model of setting OTEL_EXPORTER_OTLP_ENDPOINT on the SpinApp does it work very well with any executors that aren't ephemeral like the shim. Some executors might not be able to change there exporter endpoint on the fly like the shim can.

I would expect the executor to clearly document what it supports. Not all configuration will be supported by all executors.

Reply:
Thanks @kate-goldenring. I think that stymies my concerns.

Reply:
One other thought that comes to mind: is there any reasons we're not doing the design work in this SKIP for setting it in the SpinAppExecutor? This feels like something we'd want off the bat.

Reply:
That would be a good addition. I can add a section that details it more

Reply:
@calebschoepp in order to unblock progress on the shim and how it consumes env var config, i removed CRD design from the scope of the SKIP